Prevent kernel death on introspection.

sumitsahrawat opened the pull request:
I have updated the code relating to introspection so that it now respects the messaging spec v5.0.
Still, there is no tooltip shown on the frontend, even though I was able to confirm (using `Debug.Trace`) that the message produced by `toJSON` (from `IHaskell.IPython.Message.Writer`) had proper structure.

I searched the ipython repository for keyword `tooltip`, and it seems like it is implemented in IPython but not in jupyter. We might have to implement custom tooltips for IHaskell.

This will be required for #347.

sumitsahrawat commented:
## More details

For an example cell,

```haskell
sin pi
```

Tracing `toJSON` for `PublishDisplayData`, we get:

```json
{
    "data": {
        "text/plain": "1.2246467991473532e-16"
    },
    "metadata": {},
    "source": "haskell"
}
```

Tracing `toJSON` for `InspectReply`, with inspection on `pi` gives:

```json
{
    "status": "ok",
    "data": {
        "text/html": "<pre><b>Type:<b> forall a. Floating a => a</pre>",
        "text/plain": "forall a. Floating a => a"
    },
    "metadata": {}
}
```

According to the [spec](http://ipython.org/ipython-doc/dev/development/messaging.html#introspection), the `data` should be a mime-bundle similar to `display_data`, which it is as seen above.

sumitsahrawat commented:
@gibiansky  The ipython devs (@Carreau) on gitter told me that just sending a `inspect_reply` message should work. I have done that but I think I'm missing something. Can you take a look?

Also, introspection on the master branch causes the kernel to die. With this code, it doesn't. This seems to be a step in the right direction.

gibiansky commented:
Alright, will do!

Carreau commented:
To be more clear it should work on shift-tab, otherwise the handler won't know where to put the message.

Can be changed though.

Sent from my iPhone

> On 4 avr. 2015, at 23:23, Sumit Sahrawat <notifications@github.com> wrote:
> 
> @gibiansky The ipython devs (@Carreau) on gitter told me that just sending a inspect_reply message should work. I have done that but I think I'm missing something. Can you take a look?
> 
> Also, introspection on the master branch causes the kernel to die. With this code, it doesn't. This seems to be a step in the right direction.
> 
> —
> Reply to this email directly or view it on GitHub.
>

sumitsahrawat commented:
You mean to say that if the `inspect_reply` is sent without first receiving a `inspect_request` (sent by a shift-tab), the tooltip would not show, right?

Currently, my code doesn't work even on shift-tab, and I have been unable to find a reason for it.

Thanks for the help.

gibiansky commented:
What's the state of this PR?  Specifically:

- What exactly does it enable?
- Is it ready to merge?
- If not, what still needs to be done to get it merge-ready?

sumitsahrawat commented:
> What exactly does it enable?

Currently, this receives the input (whole cell text + cursor position), finds the word at that position (no proper lexing), and prints a lot of info, including the proper types in the console if `kernelDebug = True`. I also implemented proper replies, but the tooltip doesn't show for some unknown reason.

> Is it ready to merge?

Not yet.

> If not, what still needs to be done to get it merge-ready?

1. Proper lexing needs to be done, most probably using the GHC API.
2. The tooltip has to appear.

I tried tracing the incoming-outgoing json messages and nothing looks fishy. I also changed `replyTo` to send a fixed reply, ignoring the contents of the request and yet it didn't work.

sumitsahrawat commented:
I've edited the PR so that it prevents kernel death on introspection. The tooltip still does **not** show, but this can be merged as it stands.

gibiansky commented:
Perfect. I will investigate and merge today

sumitsahrawat commented:
Having a strange issue on travis. `./build.sh ihaskell` works for me (GHC 7.10.1). Travis shows an error in a very unexpected place.

gibiansky commented:
The build failures were unrelated I believe to this PR. Some of them may have been fixed in [this one](https://github.com/gibiansky/IHaskell/pull/488) which I just merged (I did not realize it was ready to merge). There may also have been some changes to `classy-prelude` which are breaking things, but I got things to build fine; I'll investigate travis issues separately from this PR.

Can I go ahead and merge this, then?

sumitsahrawat commented:
You can go ahead and merge this. I don't have anything else I can add.
You could also try restarting the build, now that #488 has been merged.

gibiansky commented:
@sumitsahrawat It turned out that you needed a `found` key to be true in order for the tooltip to display. This was completely undocumented (although about a week ago there was an ipython PR to document this, but the documentation has not yet appeared in the messaging spec online in HTML form). Once you added a `found` key everything displayed and your PR worked.
